newton/_src/usd/schemas.py

```diff
@@ -0,0 +1,289 @@
+
+
+def solref_to_damping(solref):
+    """Convert MuJoCo solref (timeconst, dampratio) to internal damping.
+
+    Standard mode (both positive): b = 2 * dampratio / timeconst
+    Direct mode (both negative): b = -dampratio (encodes -damping directly)
+    """
+    try:
+        timeconst = float(solref[0])
+        dampratio = float(solref[1])
+    except (TypeError, ValueError, IndexError):
+        return None
+
+    # Direct mode: both negative → solref encodes (-stiffness, -damping)
+    if timeconst < 0.0 and dampratio < 0.0:
+        return -dampratio
+
+    # Standard mode: b = 2 * dampratio / timeconst
+    if timeconst <= 0.0 or dampratio <= 0.0:
+        return None
+
+    return (2.0 * dampratio) / timeconst
+
+
+class SchemaResolverMjc(SchemaResolver):
+    name: ClassVar[str] = "mjc"
+
+    mapping: ClassVar[dict[PrimType, dict[str, SchemaAttribute]]] = {
+        PrimType.SCENE: {
+            "time_step": SchemaAttribute("mjc:option:timestep", 0.002),
+            "max_solver_iterations": SchemaAttribute("mjc:option:iterations", 100),
+            "enable_gravity": SchemaAttribute("mjc:flag:gravity", True),
+            "rigid_contact_margin": SchemaAttribute("mjc:option:o_margin", 0.0),
+        },
+        PrimType.JOINT: {
+            "armature": SchemaAttribute("mjc:armature", 0.0),
+            "friction": SchemaAttribute("mjc:frictionloss", 0.0),
+            # Per-axis linear aliases mapped to solref
+            "limit_transX_ke": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_stiffness),
+            "limit_transY_ke": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_stiffness),
+            "limit_transZ_ke": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_stiffness),
+            "limit_transX_kd": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_damping),
+            "limit_transY_kd": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_damping),
+            "limit_transZ_kd": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_damping),
+            "limit_linear_ke": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_stiffness),
+            "limit_angular_ke": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_stiffness),
+            "limit_rotX_ke": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_stiffness),
+            "limit_rotY_ke": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_stiffness),
+            "limit_rotZ_ke": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_stiffness),
+            "limit_linear_kd": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_damping),
+            "limit_angular_kd": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_damping),
+            "limit_rotX_kd": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_damping),
+            "limit_rotY_kd": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_damping),
+            "limit_rotZ_kd": SchemaAttribute("mjc:solref", [0.02, 1.0], solref_to_damping),
```

Comment:
This feels like a fairly odd way to be encoding "mujoco has one value that affects stiffness & damping for all DOFs". Is there no way to do the solref transformation once & mark the newton attrs it applies to?

Reply:
Also `MjcJointAPI` does not have `mjc:solref` at all, it has `mjc:solreffriction` and `mjc:solreflimit`